[Feat] my.dart, app_theme.dart - add my page and app theme to pages

pongpongie opened the pull request:
## #️⃣연관된 이슈

#13 

## 📝작업 내용

### `my.dart` 구현
`my.dart`는 `Header`와 `Footer`를 포함하여 `MyContent`와 `MyFloatingButton`를 Scaffold를 사용하여 레이아웃을 만들고 있습니다.

#### `my_content`
두 개의 `MyMenu`로 이뤄진 Column입니다. 푸터와 헤더를 제외한 컨텐츠들이 보여질 공간으로, 스크랩 추가 혹은 키워드 추가가 가능합니다.

#### `my_floating_button`
`FloatingActionButton`을 활용하여 액션 버튼을 구현하였습니다.
`onTap()`시에 처음 나오는 모달은 `showModalBottomSheet`를 활용하였고, 이후 각각의 메뉴를 눌러서 나오는 dialog는 `showDialog`를 사용하였습니다. `AlertDialog` 위젯을 활용하여 보여지고, 안의 내용은 `FloatModal`이라는 공용 위젯으로 만들었습니다. **이는 위젯명 수정이 필요해 보입니다.**

#### `my_menu`
`GridView`를 활용하여 스크랩 폴더와 키워드들의 view를 구현하였습니다. `onTap()`시에 `my_post.dart`로 넘어갑니다.

#### `my_post`
`ListView`를 활용하여 스크랩한 페이지들의 간단한 표시를 가능하게 만들었습니다. 리스트 뷰, 이미지 뷰 등 목록화를 추가적으로 제공하는 기능을 보완할 수 있습니다.

### `app_theme.dart`구현
**defee**의 theme을 등록하였습니다. 이에 여태 merge 되었던 페이지들에 대한 여러 값들도 theme에 맞게 수정하였습니다.

사용법 및 자세한 사항은 추가사항에서 더 설명하겠습니다.

## 💡 추가사항
### 보완사항
- `FloatModal`이라는 공용 위젯에 대한 위젯명 수정
- `my_post` image post view, list view 구현

### app_theme 사용하는법
기존에 `Color`, `TextStyle` 등의 사용이 너무 불편하고  통일되지 못해서 유지보수에 어려움이 있어서 이를 하나로 관리할 수 있는 app_theme.dart를 만들었습니다.

디자인은 material design을 참고하였으며, 추가된 theme들은 다음과 같습니다.
#### 1. DefeeColors 클래스
앱에서 사용하는 색상 상수를 정의한 클래스입니다. 앱 전체에서 일관된 색상을 사용하기 위해 사용됩니다.

- 사용법:
  원하는 색상 코드를 직접 호출하여 사용합니다.
```dart
Container(
  color: DefeeColors.blue,
  child: Text('Blue Background', style: TextStyle(color: DefeeColors.white)),
);
```
- 주요 색상:
  - blue: 기본 파란색 (#002686)
  - white: 흰색 (#FFFFFF)
  - black: 검은색 (#000000)
  - lightBlue: 연한 파란색 (#8B8FA8)
  - grey: 회색 (#BABABA)
  - surfaceContainerGrey: 밝은 회색 (#EEEDF5)
  - red: 빨간색 (#BA1A1A)

#### 2. **`DefeeThemeSizes` 클래스**
앱에서 사용하는 크기 및 여백 설정을 정의한 클래스입니다. `BorderRadius`, `Padding`, `Margin` 등의 크기를 쉽게 관리할 수 있습니다.

- 사용법:
  ```dart
  Container(
    decoration: BoxDecoration(
      borderRadius: DefeeThemeSizes.borderRadius,
    ),
    padding: DefeeThemeSizes.thinPadding,
    margin: DefeeThemeSizes.marginInsets,
  );
  ```

- 주요 설정:
  - `borderRadius`: 기본 모서리 반경 (20.0)
  - `primaryBorderRadius`: 기본의 절반 크기 (10.0)
  - `chipBorderRadius`: 기본의 두 배 크기 (40.0)
  - `padding`: 얇은 여백 (8.0)
  - `thickPadding`: 두꺼운 여백 (16.0)
  - `margin`: 기본 여백 (8.0)

#### 3. `DefeeTextStyles` 클래스
앱에서 사용하는 텍스트 스타일을 정의한 클래스입니다. 다양한 크기와 색상의 텍스트 스타일을 제공합니다.

- 사용법:
  ```dart
  Text('This is Body Text', style: DefeeTextStyles.bodyMedium);
  Text('Primary Button Text', style: DefeeTextStyles.onPrimaryMedium);
  ```

- 주요 스타일:
  - `bodyTiny`, `bodySmall`, `bodyMedium`, `bodyLarge`: 기본 텍스트 스타일 (10pt, 14pt, 18pt, 22pt)
  - `onPrimarySmall`, `onPrimaryMedium`, `onPrimaryLarge`: 밝은 배경에서 사용할 흰색 텍스트 스타일
  - `hint`, `hintSmall`: 힌트 텍스트 스타일 (회색 계열)

#### 4. `AppTheme` 클래스
앱의 테마를 관리하는 클래스입니다. 밝은 테마(`lightTheme`)와 어두운 테마(`darkTheme`)를 제공합니다. 또한, `ColorScheme`을 기반으로 앱의 기본 스타일을 구성합니다.

#### 4.1 **`lightTheme`**
앱의 밝은 테마를 정의합니다. `lightColorScheme`을 사용하여 구성됩니다.
이후에 다크모드 구현을 위해 미리 만들어 두었습니다.

#### 4.2 **`themeData` 메서드**
테마 데이터를 동적으로 생성하는 메서드입니다. `ColorScheme`을 기반으로 커스텀 테마를 만들 수 있습니다.

#### 4.4 **`lightColorScheme`**
앱의 밝은 테마 색상 팔레트입니다.

`lightColorScheme`에 정의된 색상들은 `ThemeData`를 통해 전역적으로 사용하거나, 특정 위젯에서 `Theme.of(context).colorScheme`를 통해 가져와서 사용할 수 있습니다

- `lightColorScheme` 주요 색상
  - `primary`: 주요 색상 (파란색, `#002686`)
  - `onPrimary`: 주요 색상 위에 표시될 텍스트나 아이콘 색상 (흰색, `#FFFFFF`)
  - `secondary`: 보조 색상 (연한 파란색, `#8B8FA8`)
  - `onSecondary`: 보조 색상 위에 표시될 텍스트나 아이콘 색상 (흰색, `#FFFFFF`)
  - `secondaryContainer`: 보조 색상의 밝은 버전 (연보라색, `#D1D8FF`)
  - `onSecondaryContainer`: 보조 색상 컨테이너 위의 텍스트 색상 (짙은 파란색, `#384165`)
  - `surface`: 기본 배경 색상 (아주 연한 보라색, `#FBF8FF`)
  - `onSurface`: 배경 위의 텍스트 색상 (짙은 회색, `#1A1B21`)
  - `surfaceContainer`: 서브 배경 색상 (밝은 회색, `#EEEDF5`)
  - `surfaceDim`: 반투명 배경색 (반투명 검은색, `#44121319`)
  - `outline`: 윤곽선 색상 (중간 회색, `#757683`)
  - `error`: 에러 상태 색상 (빨간색, `#BA1A1A`)
  - `onError`: 에러 상태 텍스트 색상 (흰색, `#FFFFFF`)

- `lightColorScheme` 색상 사용법
1. **앱의 전역 테마에서 사용하기**

```dart
MaterialApp(
  theme: AppTheme.lightTheme,
  home: HomePage(),
);
```

2. 위젯에서 `ColorScheme` 가져와 사용하기
**위젯 내에서 `Theme.of(context).colorScheme`로 `lightColorScheme`의 색상을 가져와 사용할 수 있습니다.**

```dart
Container(
  color: Theme.of(context).colorScheme.primary,
  child: Text(
    'Primary Color Background',
    style: TextStyle(color: Theme.of(context).colorScheme.onPrimary),
  ),
);
```
3. **버튼에 적용하기**
`ElevatedButton`, `TextButton` 등에서 `ColorScheme` 색상을 사용해 스타일링할 수 있습니다.

```dart
ElevatedButton(
  style: ElevatedButton.styleFrom(
    backgroundColor: Theme.of(context).colorScheme.secondary,
    foregroundColor: Theme.of(context).colorScheme.onSecondary,
  ),
  onPressed: () {},
  child: Text('Secondary Button'),
);
```
4. **텍스트에 적용하기**
텍스트 위젯에서 색상을 `TextStyle`에 적용할 수 있습니다. 
하지만 웬만하면 `DefeeTextStyle` 클래스로부터 사용해주세요

```dart
Text(
  'Error Message',
  style: TextStyle(
    color: Theme.of(context).colorScheme.error,
  ),
);
```

---
마지막으로 당부드리고 싶은건, 앞으로 어쨌든 색상 주거나 바꿀일 있으면 이미 제가 light theme 적용해놔서 기본 widget들 style은 적용한 대로 들어갈 겁니다. 근데 커스텀해서 사용할 때도 웬만하면 `Color.blue`이렇게 하지 말고 여기 있는 app theme에 있는 스타일 직접 사용하는 것으로 컨벤션 정하는 것이 좋아 보입니다. 

사용하실 때 `import 'package:defeefront/themes/app_theme.dart';` 하셔야 합니다.

onPrimary, primary, secondary, onSecondary 등등이 뭘 의미하는지는 다음 링크를 참고하시면 좋을 것 같습니다.
- https://m3.material.io/styles/color/roles

theme 구성은 다음 블로그를 참고하였습니다.
- https://uzzam.dev/33

**Color의 경우 context가 있는 위젯에서는 꼭 `Theme.of(context).colorScheme.onSecondary`와 같이 사용하고, 없는 경우에만 클래스 직접 사용( ex - `DefeeColors.blue`) 하는걸로 저희끼리 컨벤션 정하는 것으로 하면 좋을 것 같습니다~**

CamelIsTheBestConvention reviewed (approved):
그러면 기존에 사용되던 색상을 해당 theme의 색상을 가져와 적용하면 되나요?

pongpongie commented:
> 그러면 기존에 사용되던 색상을 해당 theme의 색상을 가져와 적용하면 되나요?

네 제가 거의 많이 바꿔놓긴 했는데, 나중에 리팩토링할 때 하나씩 수정하면 될 것 같아요
앞으로는 저 theme에서 가져다 쓰는 거로 하면 될 것 같습니다 - 이거에 대해서 컨벤션 적어놨으니 잘 읽어주세요~

Indwoo reviewed (approved):
확인했습니다~
기존에 사용하던 색깔들 수정하기

pongpongie commented:
This pull request introduces several changes to the `defeefront` project, primarily focusing on implementing a new theme system and enhancing the `MyPage` screen. The key changes include the integration of a custom theme, updates to various widgets to use the new theme, and the addition of new components for the `MyPage` screen.

### Theme Integration:
* [`defeefront/lib/main.dart`](diffhunk://#diff-70d7c510fc196e12d5cfbb7469128de2e4f611d3b91039739d1da9a41a677ea1R4-R5): Imported `AppTheme` and updated the theme settings in the `MyApp` class to use `AppTheme.lightTheme` and `AppTheme.darkTheme` with `ThemeMode.system`. [[1]](diffhunk://#diff-70d7c510fc196e12d5cfbb7469128de2e4f611d3b91039739d1da9a41a677ea1R4-R5) [[2]](diffhunk://#diff-70d7c510fc196e12d5cfbb7469128de2e4f611d3b91039739d1da9a41a677ea1L15-R19)
* Updated various widgets to use the new theme styles and colors from `AppTheme` and `DefeeTextStyles`. [[1]](diffhunk://#diff-b68b6d791ffd454ed7664b405bab566ca2cf59ce629b3bd343080e8785489a85R1) F4bf1b4bR1, F554f94bR1, [[2]](diffhunk://#diff-50d096bc0e6bad0a2e624d6253b2e59509b6deaa447b242b8ef84e78ee378922R1)

### Route Updates:
* [`defeefront/lib/routes.dart`](diffhunk://#diff-75ac1c1f36e18a75d2db5ec43295788215df3168cd9243f8c1d43b8f893ca5d7L6-R6): Replaced the `/my` route to point to the new `MyPage` widget instead of `PostPage`. [[1]](diffhunk://#diff-75ac1c1f36e18a75d2db5ec43295788215df3168cd9243f8c1d43b8f893ca5d7L6-R6) [[2]](diffhunk://#diff-75ac1c1f36e18a75d2db5ec43295788215df3168cd9243f8c1d43b8f893ca5d7L17-R17)

### New `MyPage` Screen:
* [`defeefront/lib/screens/my/my.dart`](diffhunk://#diff-ad7e8928a4172cb822a0137b6f53d1f9c17aa3a92c5975471931916952c16911R1-R51): Added a new `MyPage` widget with state management for folders and keywords, including methods for adding new folders and keywords.
* Added supporting widgets for `MyPage`:
  * `MyContent` to display folders and keywords.
  * `MyFloatingButton` for adding new folders and keywords.
  * `MyMenu` for displaying menu items.
  * `MyPost` for displaying posts within a folder.

These changes collectively enhance the user interface and user experience by introducing a consistent theme and improving the functionality of the `MyPage` screen.
